xapian-core/diversify/diversify.cc

```diff
@@ -0,0 +1,218 @@
+
+#include "debuglog.h"
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <vector>
+
+using namespace Xapian;
+using namespace std;
+
+Diversify::Diversify(unsigned int k_,
+		     double lambda_,
+		     double b_,
+		     double sigma_sqr_)
+    : k(k_), lambda(lambda_), b(b_), sigma_sqr(sigma_sqr_)
+{
+    LOGCALL_CTOR(API, "Diversify", k_ | lambda_ | b_ | sigma_sqr_);
+}
+
+string
+Diversify::get_description() const
+{
+    return "Diversify()";
+}
+
+void
+Diversify::initialise_points(const MSet &source)
+{
+    LOGCALL_VOID(API, "Diversify::initialise_points", source);
+    unsigned int count = 0;
+    TermListGroup tlg(source);
+    for (MSetIterator it = source.begin(); it != source.end(); ++it) {
+	points.emplace(*it, Xapian::Point(tlg, it.get_document()));
+	scores[*it] = it.get_weight();
+	// Initial top-k diversified documents
+	if (count < k) {
+	    main_dmset.push_back(*it);
+	    ++count;
+	}
+    }
+}
+
+pair<Xapian::docid, Xapian::docid>
+Diversify::get_key(const Xapian::docid &docid_a, const Xapian::docid &docid_b)
+{
+    pair<Xapian::docid, Xapian::docid> key;
+    if (docid_a <= docid_b) {
+	key = make_pair(docid_a, docid_b);
+    } else {
+	key = make_pair(docid_b, docid_a);
+    }
+
+    return key;
+}
+
+void
+Diversify::compute_similarities()
+{
+    LOGCALL_VOID(API, "Diversify::compute_similarities", NO_ARGS);
+    Xapian::CosineDistance d;
+    for (auto p_a : points) {
+	Xapian::docid pointid_a = p_a.first;
+	Xapian::Point point_a = p_a.second;
+	for (auto p_b : points) {
+	    Xapian::docid pointid_b = p_b.first;
+	    Xapian::Point point_b = p_b.second;
+
+	    if (pointid_a > pointid_b) {
+		continue;
+	    }
+
+	    double sim = d.similarity(point_a, point_b);
+	    auto key = get_key(pointid_a, pointid_b);
+	    pairwise_sim[key] = sim;
+	}
+    }
+}
+
+vector<Xapian::docid>
+Diversify::compute_diff_dmset(const vector<Xapian::docid> &dmset)
+{
+    LOGCALL(API, vector<Xapian::docid>, "Diversify::compute_diff_dmset", dmset);
+    vector<Xapian::docid> diff_dmset;
+    for (auto point : points) {
+	Xapian::docid point_id = point.first;
+	bool found_point = false;
+	for (auto doc_id : dmset) {
+	    if (point_id == doc_id) {
+		found_point = true;
+		break;
+	    }
+	}
+
+	if (!found_point) {
+	    diff_dmset.push_back(point_id);
+	}
+    }
+
+    return diff_dmset;
+}
+
+double
+Diversify::evaluate_dmset(const vector<Xapian::docid> &dmset)
+{
+    LOGCALL(API, double, "Diversify::evaluate_dmset", dmset);
+    double score_1 = 0, score_2 = 0;
+
+    for (auto doc_id : dmset)
+	score_1 += scores[doc_id];
+
+    vector<Xapian::docid> diff_dmset = compute_diff_dmset(dmset);
+
+    for (auto point_id : diff_dmset) {
+	double min_dist = numeric_limits<double>::max();
+	unsigned int pos = 1;
+	for (auto doc_id : dmset) {
+	    auto key = get_key(point_id, doc_id);
+	    double sim = pairwise_sim[key];
+	    double weight = 2 * b * sigma_sqr * (1 / log(1 + pos)) * (1 - sim);
```

Comment:
The compiler is fairly limited in what optimisations it can do for floating point calculations (unless you use options like GCC's `funsafe-math-optimizations` which don't conform to the language standards and so aren't on by default in sensible compilers).  So if you write:

```
2 * b * sigma_sqr * (1 / log(1 + pos)) * (1 - sim)
```

the compiler may be forced to generate an extra multiplication and use of a constant `1` compared to:

```
2 * b * sigma_sqr / log(1 + pos) * (1 - sim)
```

The second version also seems clearer to read if anything.

Reply:
I thought GCC would optimise the expression by default and thus chose to use a more verbose representation, thanks for the sharing this information. I personally feel that the latter expression requires more mental processing while keeping the left to right precedence of (*) in mind. Although, this might be because I'm not used to that.

Reply:
Both multiplication and division have the same precedence in C/C++ (I think that's true almost universally): http://en.cppreference.com/w/cpp/language/operator_precedence

You could write it as `2 * b * sigma_sqr * (1 - sim) / log(1 + pos)` if the order of evaluation doesn't matter (it clearly doesn't matter in real arithmetic, but with floating point it could matter if the intermediate values overflow or underflow the exponent range of `double` - that's unlikely to be a concern here though).

BTW, you can see the code generation difference here: https://godbolt.org/g/99cd3X

In fact, even with `-ffast-math` (which turns on `-funsafe-math-optimizations` and more) the code generated is better for: `2 * b * sigma_sqr / log(1 + pos) * (1 - sim)`

Reply:
Ah, I see. In hindsight, there was never any emphasis on such micro-level optimisations in my university courses. Thanks for pointing this out, definitely something I didn't know earlier.